File: exercism/matrix/matrix.rb
class Matrix
  def initialize(str)
    @str_matrix = str
  end

  def rows
    @rows ||= transform
  end

  def columns
    @columns ||= rows.transpose
  end

  private

  def transform
    @str_matrix.lines.map { |line| line.split.map(&:to_i) }
  end
end

#matrix = Matrix.new("9 8 7\n19 18 17")
#p matrix.rows[0], matrix.columns[0]
#p matrix.rows[1], matrix.columns[1]

File: js/exercism/matrix/matrix.js
//
// This is only a SKELETON file for the 'Matrix' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Matrix {
  constructor(string) {
    this.string = string;
  }

  // auto cashing
  get rows() {
    return this._calcRows();
  }

  get columns() {
    return this._calcColumns();
  }

  _calcRows() {
    return this.string.split('\n').map((line) => line.split(/\s/).map((c) => parseInt(c)));
  }

  _calcColumns() {
    return this._transPos(this._calcRows());
  }

  _transPos(matrix) {
    let newMatrix = [];
    for(let j=0; j < matrix[0].length; j += 1) {
      let column = [];
      for(let i=0; i < matrix.length; i += 1) {
        column.push(matrix[i][j]);
      }
      newMatrix.push(column);
    }
    return newMatrix;
  }
}
